### PR TITLE
[None][feat] Temporally-Correlated Heuristic-guided Indexer TopK for Sparse Attention

### PR DESCRIPTION
**Pull request by longcheng-nv**
## Summary

- Adds a **heuristic-guided TopK name GVR Top-K ("Guess-Verify-Refine") decode path** to the DSA (Dynamic Sparse Attention) indexer that exploits temporal correlation between consecutive decode steps, reusing previous TopK indices as candidate hints to reduce search depth
- Implements a **3-phase adaptive threshold search** with 2048-bin parallel histogram (V2e kernel with OPT5/6/7 optimizations) in a standalone compilation unit (`heuristic_topk.cuh`)
- Supports **multi-row (MTP, next_n > 1)** execution and **CUDA Graph compatibility** via per-layer metadata buffers with stable addresses
- Coexists with upstream CuTE DSL TopK in a **3-way priority dispatch**: heuristic → CuTE DSL → CUDA C++ radix sort
- **Disabled by default** (`enable_heuristic_topk=False`); opt-in via `DeepSeekSparseAttentionConfig(enable_heuristic_topk=True)`

### Commit Breakdown (11 commits)

1. **feat**: Add heuristic pre-indexed candidate path to IndexerTopK decode kernel
2. **perf**: Split heuristic TopK kernel into separate compilation unit
3. **perf**: Reuse shared heuristic TopK kernel across decode paths
4. **feat**: Extend heuristic TopK decode to multi-row (MTP) support
5. **perf**: Refactor heuristic TopK to `__noinline__` device function pattern
6. **perf**: Use adaptive `launch_bounds` and restore sort threshold
7. **feat**: Reuse previous TopK indices as heuristic hints for decode indexer
8. **fix**: Fix heuristic TopK pre_idx offset and config propagation
9. **feat**: CUDA-graph-safe heuristic TopK with per-layer metadata buffers
10. **feat**: Add DSA sparse attention and MTP support to LongBench evaluation
11. **perf**: Upgrade heuristic TopK kernel to V2e with three optimizations (OPT5/6/7)

### Key Files

| File | Description |
|------|-------------|
| `cpp/tensorrt_llm/kernels/heuristic_topk.cuh` | Core V2e kernel: 3-phase adaptive threshold with 2048-bin histogram |
| `cpp/tensorrt_llm/kernels/heuristicTopKDecode.cu` | Multi-row kernel wrapper |
| `cpp/tensorrt_llm/kernels/indexerTopK.cu` | Dispatch logic with heuristic eligibility check |
| `tensorrt_llm/_torch/attention_backend/sparse/dsa.py` | DSA backend: 3-way dispatch, per-layer CUDA Graph buffers |
| `tensorrt_llm/_torch/model_config.py` | `enable_heuristic_topk` config propagation |
| `tensorrt_llm/llmapi/llm_args.py` | `DeepSeekSparseAttentionConfig.enable_heuristic_topk` user option |

### API

Enable via `DeepSeekSparseAttentionConfig(enable_heuristic_topk=True)` in `LlmArgs`.

## Test plan

- [x] Unit tests: `pytest tests/unittest/_torch/thop/parallel/test_indexer_topk.py` (distribution-parameterized correctness tests)
- [x] Verify CUDA Graph capture/replay with heuristic path enabled
- [x] Kernel correctness: all output indices verified identical to `torch.topk` across all test configurations

#### MMLU (14,042 questions)

| | Exp 1 | Exp 2 | Exp 3 | **Average** |
|---|---|---|---|---|
| Default TRT-LLM indexer TopK | 87.51 | — | — | **87.51** |
| Heuristic-guided indexer TopK | 87.52 | 87.51 | 87.49 | **87.51** |

Delta: **0.00** — accuracy-neutral.

#### GSM8K (5-shot)

| | Exp 1 | Exp 2 | Exp 3 | **Average** |
|---|---|---|---|---|
| Default TRT-LLM indexer TopK | 95.11 | — | — | **95.11** |
| Heuristic-guided indexer TopK | 95.03 | 95.49 | 95.19 | **95.24** |

Delta: **+0.13** — accuracy-neutral (within noise).

#### GPQA-Diamond (CoT zero-shot)

| | Exp 1 | Exp 2 | Exp 3 | **Average** |
|---|---|---|---|---|
| Default TRT-LLM indexer TopK | 77.27 | — | — | **77.27** |
| Heuristic-guided indexer TopK | 75.76 | 75.25 | 78.79 | **76.60** |

Delta: **−0.67** — within GPQA-Diamond's ~3pt stderr.

#### LongBench v1 (8 runs, MTP=1)

| | Exp 1 | Exp 2 | Exp 3 | Exp 4 | Exp 5 | Exp 6 | Exp 7 | Exp 8 | **Average** |
|---|---|---|---|---|---|---|---|---|---|
| Default TRT-LLM indexer TopK | 44.37 | 45.01 | 44.68 | 44.72 | 44.47 | 44.25 | 44.64 | 44.74 | **44.61** |
| Heuristic-guided indexer TopK | 44.00 | 44.46 | 44.18 | 44.67 | 44.52 | 43.73 | 44.42 | 44.29 | **44.28** |

Delta: **−0.33** — accuracy-neutral (< 0.75%).

#### LongBench v2 (215 questions, medium context length, MTP=1)

| | Exp 1 | Exp 2 | Exp 3 | Exp 4 | Exp 5 | **Average** |
|---|---|---|---|---|---|---|
| Heuristic TopK | 49.30% | 50.23% | 48.37% | 49.77% | 47.91% | **49.12%** |
| Default TopK | 50.70% | 49.77% | 52.09% | 47.91% | 47.44% | **49.58%** |

Delta: **−0.46%** — LongBench v2 has high per-run variance (215 questions, range 47–52% even for baseline). With 5 matched runs per config, the delta is within expected noise.

#### Summary

| Benchmark | Default TopK | Heuristic TopK | Delta |
|---|---|---|---|
| **MMLU** (14K questions) | 87.51 | 87.51 | 0.00 |
| **GSM8K** (5-shot) | 95.11 | 95.24 | +0.13 |
| **GPQA-Diamond** (CoT) | 77.27 | 76.60 | −0.67 |
| **LongBench v1** (8 runs) | 44.61 | 44.28 | −0.33 |
| **LongBench v2** (medium context length, 5 runs) | 49.58% | 49.12% | −0.46% |

All benchmarks are accuracy-neutral within their respective noise margins. MMLU and GSM8K (large-N, deterministic) show effectively zero impact. GPQA and LongBench (small-N, high variance) show deltas within expected run-to-run noise.

### Performance Results (B200)

The heuristic TopK micro-kernel (`heuristicTopKMultiRowKernel`) is benchmarked against the default radix-sort path (`topKPerRowDecode`) on single-row logits input (one CTA per row). All benchmark runs first verify output indices match `torch.topk` exactly.

#### Random input (norm-similar distribution, y = 1 + 0.1 * N[0,1])

| N (seq length) | Heuristic (ns) | Radix Sort (ns) | Speedup |
|---------------:|----------------:|-----------------:|--------:|
| 8,192 | 16,512 | 11,200 | 0.68x |
| 16,384 | 21,856 | 21,984 | 1.01x |
| 32,768 | 26,112 | 32,928 | 1.26x |
| 65,536 | 31,904 | 47,936 | 1.50x |
| 70,690 | 36,864 | 51,200 | 1.39x |
| 131,072 | 43,392 | 76,128 | 1.75x |

Heuristic kernel wins at N >= 16K; at shorter sequences, the radix sort is faster (expected — the heuristic's histogram overhead dominates at small N, which is why the dispatch threshold gates it).

#### Realistic input (DeepSeek V3.2 decode logits, SWE-Bench ISL/OSL=64K/2K)

Profiled across 9 layers x 17 decode steps (N ~ 68.7K-70.7K):

| Metric | Value |
|--------|-------|
| Average speedup | **1.81x** |
| Range | 1.15x – 2.36x |
| Radix sort baseline | ~52–57 us (stable across layers) |
| Heuristic kernel | ~24–46 us (varies by layer/step) |

The heuristic kernel achieves **1.81x average speedup** over the radix sort on realistic DeepSeek V3.2 decode workloads at typical sequence lengths (~70K).

### CODE REVIEW

File: tensorrt_llm/llmapi/llm_args.py [hunk 1]
```diff
@@ -104,8 +104,7 @@ def Field(default: Any = ...,
 
 
 class CudaGraphConfig(StrictBaseModel):
-    """
-    Configuration for CUDA graphs.
+    """Configuration for CUDA graphs.
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply tensorrt_llm/llmapi/llm_args.py]: Done. The out-of-scope docstring changes in `cpp_custom_ops.py`, `model_config.py`, and `llm_args.py` have been:
1. **Reverted** from this PR (commit `916f758`)
2. **Split** into a separate PR: #12532

This PR now only contains docstrings for files directly modified by the feature (`dsa.py` and `test_dsa_indexer.py`).

File: examples/longbench/eval_longbench_v1.py [hunk 2]
```diff
@@ -134,10 +147,16 @@ def parse_arguments() -> argparse.Namespace:
...
-    # RocketKV configuration
-    parser.add_argument('--rocket_sparse',
-                        action='store_true',
-                        help='Use rocket sparse attention')
+    # Sparse attention configuration
+    sparse_group = parser.add_mutually_exclusive_group()
+    sparse_group.add_argument('--rocket_sparse',
+                              action='store_true',
+                              help='Use RocketKV sparse attention')
+    sparse_group.add_argument('--dsa_sparse',
```
[the remark this answers is not in the document]
Comment (longcheng-nv) [reply examples/longbench/eval_longbench_v1.py]: Done — added in commit 5f59b99.

New parametrized variant `heuristic_topk_mtp1` in `TestDeepSeekV32.test_fp8_blockscale`:
- Config: `(tp=8, ep=8, mtp_nextn=1, enable_heuristic_topk=True)`
- Evaluates MMLU + GSM8K accuracy with heuristic TopK enabled end-to-end through the LLM API
- Gated to Blackwell (SM >= 100) via `pytest.skip`

**Review by juney-nvidia — APPROVED** [review]
